### DISCUSSION TITLE
Resizable ArrayBuffer support

### DISCUSSION THREAD

**Discussion by mmgeorge** [wasm-bindgen/wasm-bindgen#4131] [Q&A] [answered]
Hi! I was wondering if there are any plans to take advantage of the [in-place resizable array buffers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/resizable) which appear to now be supported in all browsers. If my understanding is correct, this potentially allows for resizing the wasm heap without invalidating views into it. I.e., it would allow for removing the limitation mentioned in the doc for [TypedArray::view](https://docs.rs/js-sys/0.3.70/js_sys/struct.Float32Array.html#unsafety)? My use case is that I'd like to expose a buffer for managing various data columns (coordinates, attributes) directly to JS so that I can avoid the overhead of calling into wasm for certain code flows. 

I think it also means that the checks in the generated code for `getUint8Memory` could go away? 

https://www.proposals.es/proposals/Resizable%20and%20growable%20ArrayBuffers
https://2ality.com/2024/06/array-buffers-es2024.html#in-place-resizable-arraybuffers

**Comment by daxpedda** [comment, marked as answer]
Interesting, unfortunately after some digging I found that this requires the new [`WebAssembly.Memory.toResizableBuffer()`](https://www.w3.org/TR/2024/WD-wasm-js-api-2-20240911/#ref-for-dom-memory-toresizablebuffer) API, which isn't available in any browser yet.

But generally speaking, there are no plans, but I would probably look into it as soon as it becomes viable.

**Reply by mmgeorge** [reply]
Drats, didn't realize we're still missing the WASM memory part. Regardless, thanks for looking into this!

**Reply by guest271314** [reply]
> Interesting, unfortunately after some digging I found that this requires the new [`WebAssembly.Memory.toResizableBuffer()`](https://www.w3.org/TR/2024/WD-wasm-js-api-2-20240911/#ref-for-dom-memory-toresizablebuffer) API, which isn't available in any browser yet.
> 
> But generally speaking, there are no plans, but I would probably look into it as soon as it becomes viable.

Available in Chromium Version 148.0.7734.0 (Developer Build) (64-bit) and Firefox Nightly 150.0a1 (2026-03-13) (64-bit)

```
var kPageSize = 65536;
var wasmMemory = new WebAssembly.Memory({
  initial: 1,
  maximum: 10,
  address: "i32"
});
var toResizableBuffer = wasmMemory.toResizableBuffer();
console.log(wasmMemory.buffer.byteLength, toResizableBuffer.byteLength);
toResizableBuffer.resize(4 * kPageSize);
console.log(wasmMemory.buffer.byteLength, toResizableBuffer.byteLength);
```

See https://github.com/chromium/chromium/commit/0b5024d28ceba6396533abc69ec355dde01e25d4

**Reply by guest271314** [reply]
Safari support https://webkit.org/blog/17324/release-notes-for-safari-technology-preview-227
